Hymns/Tags/TagSheetViewModel.swift

```diff
@@ -0,0 +1,95 @@
+import SwiftUI
+import RealmSwift
+import Resolver
+
+class TagSheetViewModel: ObservableObject {
+
+    typealias Title = String
+    @Published var tags: [TagMeta]?
+    @Published var title: String = ""
+
+    let tagStore: TagStore
+    let identifier: HymnIdentifier
+    private let backgroundQueue: DispatchQueue
+    private let mainQueue: DispatchQueue
+    private let repository: HymnsRepository
+    private var disposables = Set<AnyCancellable>()
+
+    init(hymnToDisplay identifier: HymnIdentifier, tagStore: TagStore = Resolver.resolve(), hymnsRepository repository: HymnsRepository = Resolver.resolve(), mainQueue: DispatchQueue = Resolver.resolve(name: "main"), backgroundQueue: DispatchQueue = Resolver.resolve(name: "background")) {
+        self.identifier = identifier
+        self.tagStore = tagStore
+        self.repository = repository
+        self.mainQueue = mainQueue
+        self.backgroundQueue = backgroundQueue
+    }
+
+    func fetchHymn() {
+        repository
+            .getHymn(identifier)
+            .subscribe(on: backgroundQueue)
+            .receive(on: mainQueue)
+            .sink(
+                receiveValue: { [weak self] hymn in
+                    guard let self = self, let hymn = hymn, !hymn.lyrics.isEmpty else {
+                        return
+                    }
+                    let title: Title
+                    if self.identifier.hymnType == .classic {
+                        title = "Hymn \(self.identifier.hymnNumber)"
+                    } else {
+                        title = hymn.title.replacingOccurrences(of: "Hymn: ", with: "")
+                    }
+                    self.title = title
+            }).store(in: &disposables)
+    }
+
+    func fetchTags() {
+        tagStore.getTagsForHymn(hymnIdentifier: self.identifier)
+            .map({ entities -> [TagMeta] in
+                entities.map { entity -> TagMeta in
+                    return TagMeta(
+                        title: entity.tag,
+                        color: entity.tagColor)
+                }}).replaceError(with: [TagMeta]())
+            .receive(on: mainQueue)
+            .sink(receiveValue: { results in
+                self.tags = results
+            }).store(in: &disposables)
+    }
+
+    func addTag(tagName: String, tagColor: TagColor) {
+        self.tagStore.storeTag(TagEntity(hymnIdentifier: self.identifier, songTitle: self.title, tag: tagName, tagColor: tagColor))
+    }
+
+    func deleteTag(tagTitle: String, tagColor: TagColor) {
+        self.tagStore.deleteTag(primaryKey: TagEntity.createPrimaryKey(hymnIdentifier: self.identifier, tag: tagTitle), tag: tagTitle)
+    }
+}
+
+class TagMeta: Identifiable {
+
+    let title: String
+    let color: TagColor
+
+    init(title: String, color: TagColor) {
+        self.title = title
+        self.color = color
+    }
+}
+
+extension TagMeta: Hashable {
+    static func == (lhs: TagMeta, rhs: TagMeta) -> Bool {
+        lhs.title == rhs.title
```

Comment:
Should we not also compare the tag color?

Reply:
I'm not sure. It works fine for me as it is. What exactly is the advantage of doing so? Isn't this just necessary so that we can use ForEach on the array of this object? If so I don't think it is necessary since we never would ForEach just the label colors. But if there are performance benefits then we could do it.

Reply:
Well, you are implementing the "equals" function, and even though you aren't using it now, this stands to be confusing in the future. Basically here you are saying that as long as two tags have the same `title`, then they are equal. But semantically that doesn't make sense, because two tags should only be equal if they have the same title AND color.

Reply:
Ahhh, thanks good point. Thanks for the explanation. Makes more sense now. Resolved in recent commit